### PR TITLE
Read bitmap data from clipboard (HBITMAP) Unknown_Format_2

### PR DESCRIPTION
**Pull request by Nab0y**
## What does the pull request do?
When an image is on the clipboard, such as from paint.net, the await clipboard.GetFormatsAsync(); method returns three formats:
Unknown_Format_8 - A memory object containing a BITMAPINFO structure followed by the bitmap bits.
Unknown_Format_17 - A memory object containing a BITMAPV5HEADER structure followed by the bitmap color space information and the bitmap bits.
Unknown_Format_2 - A handle to a bitmap (HBITMAP).
[https://docs.microsoft.com/](https://docs.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats)
The current implementation reads data for structures Unknown_Format_8 and Unknown_Format_17 but **Unknown_Format_2** HBITMAP is a handle that is located in a different area of memory and it is impossible to get it in the current implementation.

## What is the current behavior?
`var imagePixelsData = await clipboard.GetDataAsync("Unknown_Format_2") as byte[]; `
throw exception

## What is the updated/expected behavior with this PR?
`var imagePixelsData = await clipboard.GetDataAsync("Unknown_Format_2") as byte[]; `
return raw image bytes

## How was the solution implemented (if it's not obvious)?
Implemented correct reading of HBITMAP and through the call UnmanagedMethods.GetBitmapBits reading bytes of the image

You can use the method for check
`private async Task ReadClipboardImageAsync()
        {
            var clipboard = Application.Current.Clipboard;

            var formats = await clipboard.GetFormatsAsync();
            if (formats.Contains("Unknown_Format_8") && formats.Contains("Unknown_Format_2"))
            {
                var imagePixelsData = await clipboard.GetDataAsync("Unknown_Format_2") as byte[];

                var bitmapInfoData = await clipboard.GetDataAsync("Unknown_Format_8") as byte[];
                var header = ReadImageHeader(bitmapInfoData);

                var pixels = new SKColor[header.biWidth * header.biHeight];
                var pixelIndex = 0;

                var pos = 0;
                for (var y = 0; y < header.biHeight; y++)
                {
                    for (var x = 0; x < header.biWidth; x++)
                    {
                        var b = imagePixelsData[pos++];
                        var g = imagePixelsData[pos++];
                        var r = imagePixelsData[pos++];
                        if (header.biBitCount == 32)
                        {
                            var a = imagePixelsData[pos++];
                            pixels[pixelIndex++] = new SKColor(r, g, b, a);
                            continue;
                        }
                        pixels[pixelIndex++] = new SKColor(r, g, b);
                    }
                }

                var bitmap = new SKBitmap(new SKImageInfo(header.biWidth, header.biHeight));
                bitmap.Pixels = pixels;

                using var writeStream = File.Open("C:\\Projects\\Sample.png", FileMode.OpenOrCreate, FileAccess.Write);
                bitmap.Encode(SKEncodedImageFormat.Png, 100).SaveTo(writeStream);
            }
        }`

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
I think no

## Obsoletions / Deprecations
no

## Fixed issues
Partial #3588

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=6501) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=6501) before we can accept your contribution.<br/><br/>:x: Nab0y [sign now](https://cla.dotnetfoundation.org/AvaloniaUI/Avalonia?pullRequest=6501)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/AvaloniaUI/Avalonia?pullRequest=6501) it.</sub>

**Comment by kekekeks** [comment]
Are you sure that we want `Unknown_format_2` instead of `CF_BITMAP` or `Win32_CF_BITMAP`? Same with `CF_DIB` and `CF_DIBV5`

**Comment by kekekeks** [comment]
I guess we are misusing [GetClipboardFormatName](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getclipboardformatnamew)

`The type of format to be retrieved. This parameter must not specify any of the predefined clipboard formats.`

So we need to hardcode those predefined format names.

**Comment by Nab0y** [comment]
> Are you sure that we want `Unknown_format_2` instead of `CF_BITMAP` or `Win32_CF_BITMAP`? Same with `CF_DIB` and `CF_DIBV5`

I worked with the formats that come back to me in the current reality.

File: src/Windows/Avalonia.Win32/OleDataObject.cs [hunk 1]
```diff
@@ -44,9 +44,77 @@ public IEnumerable<string> GetFileNames()
 
         public object Get(string dataFormat)
         {
+            if (dataFormat == ClipboardFormats.GetFormat(2)) // check for CF_BITMAP
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/OleDataObject.cs:47]: I think that to improve the reading of the code it is better to define a constant.
```csharp
const short CF_BITMAP = 2
```
end using

```csharp
if (dataFormat == ClipboardFormats.GetFormat(CF_BITMAP))
```

File: src/Windows/Avalonia.Win32/OleDataObject.cs [hunk 2]
```diff
@@ -44,9 +44,77 @@ public IEnumerable<string> GetFileNames()
...
+            if (bitmapInfoData == null)
+            {
+                return null;
+            }
+
+            var bitmapInfo = ReadBitmapInfo(bitmapInfoData);
+
+            // get bitmap handle HBITMAP
+            var formatEtc = new FORMATETC();
+            formatEtc.cfFormat = 2; // CF_BITMAP
```
Comment (workgroupengineering) [inline src/Windows/Avalonia.Win32/OleDataObject.cs:69]: same of above

File: src/Windows/Avalonia.Win32/OleDataObject.cs (hunk 1 above)
Comment (Nab0y) [reply src/Windows/Avalonia.Win32/OleDataObject.cs]: I added all the standard formats as constants from the [documentation](https://docs.microsoft.com/en-us/windows/win32/dataxchg/standard-clipboard-formats)

**Comment by Nab0y** [comment]
Sorry, but what are we waiting for?

File: src/Windows/Avalonia.Win32/OleDataObject.cs [hunk 3]
```diff
@@ -44,9 +44,77 @@ public IEnumerable<string> GetFileNames()
...
+                return GetImage();
+            }
+
             return GetDataFromOleHGLOBAL(dataFormat, DVASPECT.DVASPECT_CONTENT);
         }
 
+        private object GetImage()
+        {
+            // read bitmapInfo
+            var bitmapInfoData = GetDataFromOleHGLOBAL(ClipboardFormats.GetFormat(StandardClipboardFormats.CF_DIB), DVASPECT.DVASPECT_CONTENT) as byte[];
```
Comment (kekekeks) [inline src/Windows/Avalonia.Win32/OleDataObject.cs:58]: You are checking for the format to be CF_BITMAP and then immediately try to read it as CF_DIB. Are you sure that both formats are guaranteed to be present?

File: src/Windows/Avalonia.Win32/OleDataObject.cs [hunk 4]
```diff
@@ -44,9 +44,77 @@ public IEnumerable<string> GetFileNames()
...
+            {
+                _wrapped.GetData(ref formatEtc, out STGMEDIUM medium);
+                try
+                {
+                    if (medium.unionmember != IntPtr.Zero && medium.tymed == TYMED.TYMED_GDI)
+                    {
+                        // get image bytes
+                        var bufer = new byte[bitmapInfo.biSizeImage];
+                        UnmanagedMethods.GetBitmapBits(medium.unionmember, bufer.Length, bufer);
+                        return bufer;
```
Comment (kekekeks) [inline src/Windows/Avalonia.Win32/OleDataObject.cs:84]: Are you sure that returning a byte array with raw bitmap data but without pixel format and layout is OK? How does it work in WPF?

**Comment by jmacato** [comment]
Closing this due to inactivity and unresolved merge conflicts. Please ping me if you want to work on this again.

**Pull request closed without merging**